src/lib/Types/Core.hs

```diff
@@ -217,6 +217,46 @@ instance GenericB EffectBinder where
   fromB (EffectBinder effs) = LiftB effs
   toB   (LiftB effs) = EffectBinder effs
 
+-- === type classes ===
+
+data ClassDef (n::S) where
+  ClassDef
+    :: SourceName
+    -> [SourceName]       -- method source names
+    -> Nest Binder n1 n2  -- parameters
+    ->   [Type n2]        -- superclasses
+    ->   [MethodType n2]  -- method types
+    -> ClassDef n1
+
+data InstanceDef (n::S) where
+  InstanceDef
+    :: ClassName n1
+    -> Nest PiBinder n1 n2 -- parameters (types and dictionaries)
+    ->   [Type n2]         -- class parameters
+    ->   InstanceBody n2
+    -> InstanceDef n1
+
+data InstanceBody (n::S) =
+  InstanceBody
+    [Atom n]   -- superclasses dicts
+    [Block n]  -- method definitions
+  deriving (Show, Generic)
+
+data MethodType (n::S) =
+  MethodType
+    [Bool]    -- indicates explicit args
```

Comment:
That's true. But at the top level, source names for methods point to this piece of data. It's similar to how we keep the implicit-vs-explicit arrow types in the core IR, even though they aren't relevant after type inference. At least in this case the data only lives at the top level, so it's not like we're paying the cost of carrying the `[Bool]` in the expressions we're compiling.

Reply:
Do they have to point to this, instead of to a fully formed atom that is a lambda with a single `ProjectMethod` op inside? That would be closer to the old representation and it was nice in that we only had to elaborate the atom once, as opposed to recreating it multiple times every time it's used. WDYT?

Reply:
Method source names actually point to both. The `Binding` case looks like: `MethodBinding :: Name ClassNameC n -> Int -> Atom n -> Binding MethodNameC  n`. The atom there is the projection, so we don't have to elaborate it every time.